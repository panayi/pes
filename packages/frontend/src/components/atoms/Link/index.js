import * as R from 'ramda';
import classNames from 'classnames';
import Button from 'material-ui/Button';
import IconButton from 'material-ui/IconButton';
import { withStyles } from 'material-ui/styles';
import { NavLink } from 'react-router-dom';
import { defaultProps, withProps } from 'recompose';
import omitProps from 'utils/omitProps';

const DEFAULT_ACTIVE_CLASS_NAME = 'link--active';

const styles = {
  root: {
    textTransform: 'none',

    [`&.${DEFAULT_ACTIVE_CLASS_NAME}`]: {
      // Styles for active link go here
    },
  },
};

const makeLink = R.compose(
  defaultProps({
    activeClassName: DEFAULT_ACTIVE_CLASS_NAME,
  }),
  withStyles(styles),
  withProps(({ to, classes, className }) => ({
    component: to ? NavLink : null,
    className: classNames(classes.root, className),
  })),
  omitProps(['classes', 'dispatch']),
);

const Link = makeLink(Button);

Link.icon = makeLink(IconButton);

export default Link;
