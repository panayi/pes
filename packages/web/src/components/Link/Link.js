import * as R from 'ramda';
import classNames from 'classnames';
import IconButton from 'material-ui/IconButton';
import withStyles from 'material-ui/styles/withStyles';
import { NavLink } from 'react-router-dom';
import { defaultProps, withProps } from 'recompose';
import omitProps from 'utils/omitProps';
import Button from 'components/Button/Button';

const DEFAULT_ACTIVE_CLASS_NAME = 'link--active';

const styles = {
  root: {
    textTransform: 'none',
    cursor: 'default',

    [`&.${DEFAULT_ACTIVE_CLASS_NAME}`]: {
      // Styles for active link go here
    },
  },
  clickable: {
    cursor: 'pointer',
  },
};

const makeLink = R.compose(
  defaultProps({
    activeClassName: DEFAULT_ACTIVE_CLASS_NAME,
  }),
  withStyles(styles),
  withProps(({ to, classes, className }) => ({
    component: to ? NavLink : null,
    className: classNames(classes.root, className, {
      [classes.clickable]: !!to,
    }),
  })),
  omitProps(['classes', 'dispatch']),
);

const Link = makeLink(Button);

Link.icon = makeLink(IconButton);

export default Link;
