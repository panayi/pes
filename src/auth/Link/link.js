import * as R from 'ramda';
import { createAction, handleAction } from 'redux-actions';
import { profileEmailSelector } from '../auth.selectors';

// ------------------------------------
// Constants
// ------------------------------------
const RECEIVE_LINKED_ACCOUNTS = 'auth/link/RECEIVE_LINKED_ACCOUNTS';

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = null;

const linkedAccountsReducer = handleAction(
  [RECEIVE_LINKED_ACCOUNTS],
  (state, { payload }) => payload,
  initialState,
);

export default linkedAccountsReducer;

// ------------------------------------
// Selectors
// ------------------------------------

// linkedAccountsSelector :: State -> Array | Nil
export const linkedAccountsSelector = R.path(['auth', 'linkedAccounts']);

// ------------------------------------
// Actions
// ------------------------------------
const receiveLinkedAccounts = createAction(RECEIVE_LINKED_ACCOUNTS);

export const fetchLinkedAccounts = () => (dispatch, getState, getFirebase) => {
  const email = profileEmailSelector(getState());

  if (R.isNil(email)) {
    return dispatch(receiveLinkedAccounts([]));
  }

  return getFirebase().auth().fetchProvidersForEmail(email)
    .then(linkedAccounts => dispatch(receiveLinkedAccounts(linkedAccounts)));
};

export const actions = {
  fetchLinkedAccounts,
};
