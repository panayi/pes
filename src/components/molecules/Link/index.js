import * as R from 'ramda';
import classNames from 'classnames';
import { Button, IconButton, withStyles } from 'material-ui';
import { NavLink } from 'react-router-dom';
import { defaultProps, withProps } from 'recompose';
import omitProps from 'utils/omitProps';

const DEFAULT_ACTIVE_CLASS_NAME = 'link--active';

const styles = {
  button: {
    [`&.${DEFAULT_ACTIVE_CLASS_NAME}`]: {
      color: 'red',
    },
  },
};

const makeLink = R.compose(
  defaultProps({
    activeClassName: DEFAULT_ACTIVE_CLASS_NAME,
  }),
  withStyles(styles),
  withProps(({ to, classes, className }) => ({
    component: to ? NavLink : null,
    className: classNames(className, classes.button),
  })),
  omitProps(['classes', 'dispatch']),
);

const Link = makeLink(Button);

Link.icon = makeLink(IconButton);

export default Link;
