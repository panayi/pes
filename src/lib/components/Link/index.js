import * as R from 'ramda';
import classNames from 'classnames';
import { Button, withStyles } from 'material-ui';
import { NavLink } from 'react-router-dom';
import { defaultProps, withProps } from 'recompose';
import omitProps from '../../helpers/omitProps';

const DEFAULT_ACTIVE_CLASS_NAME = 'link--active';

const styles = {
  button: {
    [`&.${DEFAULT_ACTIVE_CLASS_NAME}`]: {
      color: 'red',
    },
  },
};

export default R.compose(
  defaultProps({
    activeClassName: DEFAULT_ACTIVE_CLASS_NAME,
  }),
  withStyles(styles),
  withProps(({ to, classes, className }) => ({
    component: to ? NavLink : null,
    className: classNames(className, classes.button),
  })),
  omitProps(['classes', 'dispatch']),
)(Button);
