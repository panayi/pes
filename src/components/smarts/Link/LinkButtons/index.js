import React from 'react';
import Link from '../../Link';

export const LinkButtons = () => (
  <Link>
    <Link.Button withProvider="google" />
    <Link.Button withProvider="facebook" />
  </Link>
);

export default LinkButtons;
