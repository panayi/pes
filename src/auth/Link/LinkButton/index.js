/* @flow */
import React, { Component } from 'react';
import * as R from 'ramda';
import capitalize from 'lodash.capitalize';
import { withProps } from 'recompose';
import { firebaseConnect } from 'react-redux-firebase';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Button } from 'rebass';
import createAuthProvider from '../../helpers/createAuthProvider';
import { auth as authConfig } from '../../../lib/config';
import { actions, linkedAccountsSelector } from '../link';
import { actions as authActions } from '../../auth';

type Props = {
  getProvider: Function,
  withProvider: String,
  providerLabel: String,
  firebase: Object,
  fetchLinkedAccounts: Function,
  updateProfile: Function,
  disabled: Boolean,
  linkedAccounts: Array<String>,
};

export class LinkButton extends Component<Props> {
  handleClick = () => {
    const { firebase, withProvider, fetchLinkedAccounts, updateProfile } = this.props;
    const provider = createAuthProvider(
      firebase,
      withProvider,
      authConfig[withProvider].scopes,
    );

    firebase.auth().currentUser
      .linkWithPopup(provider)
      .then((result) => {
        updateProfile(result.user);
        fetchLinkedAccounts();
      });
  }

  render() {
    const { providerLabel, disabled } = this.props;

    return (
      <Button
        onClick={this.handleClick}
        disabled={disabled}
      >
        Link with {providerLabel}
      </Button>
    );
  }
}

const mapToLinkedAccount = {
  facebook: 'facebook.com',
  google: 'google.com',
};

const mapStateToProps = createStructuredSelector({
  linkedAccounts: linkedAccountsSelector,
});

const mapDispatchToProps = {
  fetchLinkedAccounts: actions.fetchLinkedAccounts,
  updateProfile: authActions.updateProfile,
};

const isDisabled = R.converge(R.contains, [
  R.compose(
    R.prop(R.__, mapToLinkedAccount),
    R.prop('withProvider'),
  ),
  R.compose(
    R.defaultTo([]),
    R.prop('linkedAccounts'),
  ),
]);

export default R.compose(
  firebaseConnect(),
  connect(mapStateToProps, mapDispatchToProps),
  withProps(props => ({
    disabled: isDisabled(props),
    providerLabel: capitalize(props.withProvider),
  })),
)(LinkButton);
