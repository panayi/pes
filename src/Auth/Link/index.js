/* @flow */
import React, { Component } from 'react';
import R from 'ramda';
import { isLoaded } from 'react-redux-firebase';
import { branch, renderNothing } from 'recompose';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { profileSelector } from '../auth.selectors';
import { actions, linkedAccountsSelector } from './link';
import LinkButton from './LinkButton';

type Props = {
  children: React$Node,
  fetchLinkedAccounts: Function,
  linkedAccounts: Array<String>,
};

export class Link extends Component<Props> {
  componentWillMount() {
    this.props.fetchLinkedAccounts();
  }

  render() {
    if (R.isNil(this.props.linkedAccounts)) {
      return null;
    }

    return (
      <div>
        {this.props.children}
      </div>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  profile: profileSelector,
  linkedAccounts: linkedAccountsSelector,
});

const mapDispatchToProps = {
  fetchLinkedAccounts: actions.fetchLinkedAccounts,
};

const ConnectedLink = R.compose(
  connect(mapStateToProps, mapDispatchToProps),
  // Render only when `profile` has been fetched,
  // as the profile.email is needed for fetching linked-accounts.
  branch(
    R.compose(
      R.not,
      isLoaded,
      R.prop('profile'),
    ),
    renderNothing,
  ),
)(Link);

ConnectedLink.Button = LinkButton;

export default ConnectedLink;
