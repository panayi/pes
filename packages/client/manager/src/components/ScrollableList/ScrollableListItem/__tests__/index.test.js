import React from 'react';
import { withRouter } from 'react-router-dom';
import ListItem from '@material-ui/core/ListItem';
import Link from '@pesposa/client-core/src/components/Link/Link';
import ScrollableListItemHoc, { ScrollableListItem } from '..';

const mockStore = {
  router: {
    location: {
      pathname: 'test_path',
      search: 'test_search',
      hash: 'test_hash',
      key: 'test_key',
    },
  },
};

describe('ScrollableList', () => {
  describe('Item', () => {
    it('should render properly', () => {
      const wrapper = withMockStore(<ScrollableListItemHoc />, mockStore);
      const Component = shallow(wrapper.component);

      expect(Component).toMatchSnapshot();
    });

    it('should render as active iff this links to the current path', () => {
      const wrapperActive = withRouter(
        withMockStore(<ScrollableListItemHoc to="test_path" />, mockStore)
          .component,
      );
      const ComponentActive = mount(wrapperActive);

      const wrapperInActive = withRouter(
        withMockStore(<ScrollableListItemHoc to="other_path" />, mockStore)
          .component,
      );
      const ComponentInActive = mount(wrapperInActive);

      expect(ComponentActive.find(ScrollableListItem).prop('isActive')).toEqual(
        true,
      );
      expect(
        ComponentInActive.find(ScrollableListItem).prop('isActive'),
      ).toEqual(false);
    });

    it('should render primary & secondary texts somewhere', () => {
      const textPrimary = 'test_primary';
      const textSecondary = 'test_secondary';
      const wrapper = withMockStore(
        <ScrollableListItemHoc
          primary={textPrimary}
          secondary={textSecondary}
        />,
        mockStore,
      );
      const Component = mount(wrapper.component);

      expect(Component.text()).toContain(textPrimary);
      expect(Component.text()).toContain(textSecondary);
    });

    it('should render icon somewhere', () => {
      const iconClassName = 'testIcon';
      const icon = <span className={iconClassName} />;
      const wrapper = withMockStore(
        <ScrollableListItemHoc icon={icon} />,
        mockStore,
      );
      const Component = mount(wrapper.component);

      expect(Component.find(`.${iconClassName}`).length).not.toEqual(0);
    });

    it('should render a `ListItemSecondaryAction` iff `action` is passed', () => {
      const actionClassName = 'testAction';
      const action = <span className={actionClassName} />;
      const wrapper = withMockStore(
        <ScrollableListItemHoc action={action} />,
        mockStore,
      );
      const Component = mount(wrapper.component);

      expect(Component.find(`.${actionClassName}`).length).not.toEqual(0);
    });

    it('should pass className & style props to the ListItem root', () => {
      const className = 'testClass';
      const style = {
        border: 'none',
        color: 'inherit',
      };
      const wrapper = withMockStore(
        <ScrollableListItemHoc className={className} style={style} />,
        mockStore,
      );
      const Item = mount(wrapper.component).find(`.${className}`);

      expect(Item.length).not.toEqual(0);
      expect(Item.first().prop('style')).toEqual(style);
    });

    it('should render as a Link if `to` prop is passed', () => {
      const wrapper = withRouter(
        withMockStore(<ScrollableListItemHoc to="some_url" />, mockStore)
          .component,
      );
      const Component = mount(wrapper);

      expect(Component.find(ListItem).prop('component')).toEqual(Link);
    });

    it('should handle the passed `onClick` function', () => {
      const onClickSpy = jest.fn();
      const wrapper = withMockStore(
        <ScrollableListItemHoc onClick={onClickSpy} />,
        mockStore,
      );
      const Component = mount(wrapper.component);

      expect(onClickSpy.mock.calls.length).toEqual(0);

      Component.find(ListItem).simulate('click');

      expect(onClickSpy.mock.calls.length).toEqual(1);
    });
  });
});
