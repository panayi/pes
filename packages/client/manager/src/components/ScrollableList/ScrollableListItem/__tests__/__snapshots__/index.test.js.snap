// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollableList Item should render properly 1`] = `<Connect(WithStyles(Component)) />`;
